values/zhash/value.go

```diff
@@ -0,0 +1,288 @@
+	typ := fmt.Sprintf("%T", e)
+	return fmt.Errorf("%w: %#v not a(n) %q", ErrInvalidParamType, p, typ)
+}
+
+// newParamValueErr creates an error where parameter value is wrong.
+func newParamValueErr(v interface{}) error {
+	typ := fmt.Sprintf("%T", v)
+	return fmt.Errorf("%w: cannot process function on %q", ErrInvalidParamValue, typ)
+}
+
+const (
+	// Insert inserts the element(s) in the zhash.
+	// Default score is 0.
+	// If element is already present, updates the score to 0.
+	//
+	// Returns an array of added elements.
+	Insert kiwi.Action = "INSERT"
+
+	// Remove removes the element(s) from the zhash.
+	//
+	// Returns an array of removed elements.
+	Remove kiwi.Action = "REMOVE"
+
+	// Increment increments the score by given value.
+	//
+	// Returns updated score.
+	Increment kiwi.Action = "INCREMENT"
+
+	// Len gets the length of the zhash.
+	//
+	// Returns an integer.
+	Len kiwi.Action = "LEN"
+
+	// Get gets the score of given element.
+	//
+	// Returns an integer.
+	Get kiwi.Action = "GET"
+
+	// PeekMax gets the element with maximum score.
+	//
+	// Returns string value.
+	// Returns nil if zhash is empty.
+	PeekMax kiwi.Action = "PEEKMAX"
+
+	// PeekMin gets the element with minimum score.
+	//
+	// Returns string value.
+	// Returns nil if zhash is empty.
+	PeekMin kiwi.Action = "PEEKMIN"
+)
+
+// Type returns v's type, i.e., "zhash".
+func (v *Value) Type() kiwi.ValueType {
+	return Type
+}
+
+// DoMap returns the map of v's actions with it's do functions.
+func (v *Value) DoMap() map[kiwi.Action]kiwi.DoFunc {
+	return map[kiwi.Action]kiwi.DoFunc{
+		Insert:    v.insert,
+		Remove:    v.remove,
+		Increment: v.increment,
+		Len:       v.len,
+		Get:       v.get,
+		PeekMax:   v.peekmax,
+		PeekMin:   v.peekmin,
+	}
+}
+
+// ToJSON returns the raw byte array of v's data
+func (v *Value) ToJSON() (json.RawMessage, error) {
+	nodes := v.GetByRankRange(1, -1, false)
+
+	vals := map[string]Item{}
+	for _, node := range nodes {
+		vals[node.Key()] = Item{fmt.Sprintf("%v", node.Value), int(node.Score())}
+	}
+
+	c, err := json.Marshal(vals)
+	if err != nil {
+		return nil, err
+	}
+	return json.RawMessage(c), nil
+}
+
+// FromJSON populates the s with the data from RawMessage
+func (v *Value) FromJSON(rawmessage json.RawMessage) error {
+	vals := map[string]Item{}
+	if err := json.Unmarshal(rawmessage, &vals); err != nil {
+		return err
+	}
+
+	for key, item := range vals {
+		v.AddOrUpdate(key, sortedset.SCORE(item.Score), item.Value)
+	}
+
+	return nil
+}
+
+// insert implements the INSERT action.
+func (v *Value) insert(params ...interface{}) (interface{}, error) {
+	if len(params) < 1 || len(params) > 2 {
+		return nil, newParamLenErr(len(params), 1)
+	}
+
+	var (
+		ok    bool
+		key   string
+		value string
+	)
+
+	key, ok = params[0].(string)
+	if !ok {
+		return nil, newParamTypeErr(params[0], key)
+	}
+
+	if len(params) == 1 {
+		(*v).AddOrUpdate(key, 0, nil)
+		return key, nil
+	}
```

Comment:
This case is already taken care above.
If number of parameters passed =1, we will send error of inappropriate instead of storing `nil` value.
So this check can be removed from here.

Reply:
Ok. I did that way because @vrongmeal said this in the issue:

> When inserting, you can do both the things. Parameters are dynamic. So if you get one parameter you can just insert it, if you get two, you can insert while setting the value. There will always be a "SET" action if someone wants to insert the key and then set the value.

Reply:
It seems reasonable. Ok, then if 1 parameter is passed, value would be nil and if 2 parameters are passed value would be inserted.
But make an additional action `SET` which will take 2 parameters key and value and set the given value for the key.

Reply:
Also add the corresponding test for this action:)